chaosgcp/iam/controls/policy.py

```diff
@@ -0,0 +1,304 @@
+
+logger = logging.getLogger("chaostoolkit")
+
+__all__ = ["before_experiment_control", "after_experiment_control"]
+
+set_current_time = datetime.datetime.now(timezone.utc)
+
+
+def before_experiment_control(
+    context: Experiment,
+    project_id: str,
+    roles: List,
+    members: List,
+    configuration: Configuration = None,
+    secrets: Secrets = None,
+    iam_propogation_sleep_time_in_minutes=2,
+    expiry_time_in_minutes=10,
+) -> None:
+    """
+    to ADD IAM roles
+    """
+    # Generate expiration timestamp (current time + expiration_minutes)
+    expire_time_minutes = (
+        iam_propogation_sleep_time_in_minutes + expiry_time_in_minutes
+    )
+    expiration_timestamp = set_current_time + datetime.timedelta(
+        minutes=expire_time_minutes
+    )
+    formatted_timestamp = expiration_timestamp.strftime("%Y-%m-%dT%H:%M:%SZ")
+    roles_iam_condition = {
+        "title": "chaostoolkit",
+        "description": "Should expire soon",
+        "expression": f'request.time < timestamp("{formatted_timestamp}")',
+    }
+    try:
+        manage_time_bound_temporary_iam_roles(
+            project_id,
+            members,
+            roles,
+            roles_iam_condition,
+            "add",
+            iam_propogation_sleep_time_in_minutes,
+        )
+    except Exception as e:
+        logger.error(e, exc_info=True)
+        raise InterruptExecution(
+            "Failed to add IAM roles, Please check logs"
+        ) from e
+
+
+def after_experiment_control(
+    context: Experiment,
+    state: Journal,
+    project_id: str,
+    roles: List,
+    members: List,
+    iam_propogation_sleep_time_in_minutes=2,
+    expiry_time_in_minutes=10,
+    configuration: Configuration = None,
+    secrets: Secrets = None,
+) -> None:
+    """
+    to remove IAM roles
+    """
+    # Generate expiration timestamp (current time + expiration_minutes)
+    expire_time_minutes = (
+        iam_propogation_sleep_time_in_minutes + expiry_time_in_minutes
+    )
+    expiration_timestamp = set_current_time + datetime.timedelta(
+        minutes=expire_time_minutes
+    )
+    formatted_timestamp = expiration_timestamp.strftime("%Y-%m-%dT%H:%M:%SZ")
+    roles_iam_condition = {
+        "title": "chaostoolkit",
+        "description": "Should expire soon",
+        "expression": f'request.time < timestamp("{formatted_timestamp}")',
+    }
+    try:
+        manage_time_bound_temporary_iam_roles(
+            project_id,
+            members,
+            roles,
+            roles_iam_condition,
+            "remove",
+            iam_propogation_sleep_time_in_minutes,
+        )
+    except Exception as e:
+        logger.error(e, exc_info=True)
+        raise InterruptExecution(
+            "Failed to remove IAM roles, Please check logs"
+        ) from e
+
+
+def manage_time_bound_temporary_iam_roles(
+    project_id: str,
+    members: list[str],
+    roles: list[str],
+    roles_iam_condition: dict[str, str],
```

Comment:
Please switch to `Dict` instead of `dict`. The latter really only came into play with Python 3.9 but we support 3.8.

https://docs.python.org/3.12/library/typing.html#typing.Dict

Reply:
changed to Dict